proposals/generate_validating_admission_policies.md

```diff
@@ -0,0 +1,474 @@
+
+- When a validating admission policy or validating admission policy binding event happens, the policy key is deduced from owner references and it's enqueued. The controller will reset the changes done in either VAP or VAP binding.
+
+- A flag/toggle will be introduced in the controller so that user can either let Kyverno engine handles CEL expressions or let api server handles it by generating VAPs and their bindings.
+
+### Kyverno engine VS API server
+Since CEL expressions can be handled by either Kyverno engine or the API server, Here are the cases where Kyverno engine will handle the resource validations itself:
+1. The flag is disabled in the `validatingadmissionpolicy-controller`. This is done by setting `--generateValidatingAdmissionPolicy` to false.
+2. The flag is enabled and validating admission policies were generated but issues are encountered in processing these policies. This is done by getting the metrics exposed by the API server.
+
+### Limitations
+Validating admission policy and its bindings will be generated for a Kyverno policy that has one `validate.CEL` subrule.
+
+## Implementation
+In this section, we will discuss how validating admission policy and its binding can be created from Kyverno policies.
+
+### ClusterPolicy vs Policy
+All Kyverno policies of kind `ClusterPolicy` can be converted into validating admission policies but not all policies of kind `Policy` can be converted into validating admission policies since they are cluster-scoped resources.
+
+Kyverno policies of kind `Policy` can be converted into validating admission policy in case it contains namespace selectors as follows:
+```
+apiVersion: kyverno.io/v1
+kind: Policy
+metadata:
+  name: pol-1
+  namespaceSelector: 
+    matchExpressions:
+      - key: environment
+        operator: In
+        values:
+          - prod
+```
+It will be converted into validating admission policy binding as follows:
+```
+spec:
+  matchResources: 
+    namespaceSelector: 
+      matchExpressions:
+        - key: environment
+          operator: In
+          values:
+            - prod
+```
+
+### Kyverno Policy Settings
+In Kyverno policies, some common settings can be applied to all rules. We will discuss the effect of setting these values on generating validating admission policies assuming that Kyverno policy contains multiple rules, one of them is `validate.CEL` subrule.
+
+1. `applyRules`: 
+    - If it's set to `All` (default), Validating admission policy and its binding will be generated as a result of having `validate.CEL` subrule. 
+    - If it's set to `One` and `validate.CEL` subrule is the first matching rule, then Validating admission policy and its binding will be generated.
+    - If it's set to `One` and `validate.CEL` subrule isn't the first matching rule, no validating admission policy will be generated.
```

Comment:
I don't think this will provide the intended behaviors.

Rules are an ordered set in a Kyverno policy, and `applyRules` controls when to stop processing of rules.

It may be best to limit the generation of Validating Admission Policies to policies that contain a single rule, for the initial release.